### PR TITLE
KAFKA-15351: Ensure log-start-offset not updated to local-log-start-offset when remote storage enabled

### PR DESCRIPTION
**Pull request by kamalcph**
Part-1: When tiered storage is enabled on the topic, and the last-standing-replica is restarted, then the log-start-offset should not reset its offset to first-local-log-segment-base-offset.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by kamalcph** [comment]
@showuon @divijvaidya @clolov @satishd 

In this patch, we have covered the case where the log-start-offset gets reseted to local-log-start-offset when the last standing replica is restarted and enabled the OffloadAndConsumeFromLeaderTest with broker bounce. In the next patch, we can cover the case where the log-start-offset should be updated by the remote-log-segment deleter threads on becoming the leader. Let me know if we want to merge both the cases in the same PR.

Please take a look when you get chance!

**Comment by kamalcph** [comment]
### Test Report

```scala
[SUCCESS] (1) create topic: Topic[name=topicA partition-count=1 replication-factor=1 segment-size=1 assignment=null properties={remote.storage.enable=true, local.retention.bytes=1, index.interval.bytes=1, segment.index.bytes=12}]

[SUCCESS] (2) produce-records: topicA-0
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k1, value=v1, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k2, value=v2, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k3, value=v3, timestamp=null)
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=0 record-count=1]
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=1 record-count=1]

[SUCCESS] (3) create topic: Topic[name=topicB partition-count=1 replication-factor=1 segment-size=2 assignment=null properties={remote.storage.enable=true, local.retention.bytes=1, index.interval.bytes=1, segment.index.bytes=24}]

[SUCCESS] (4) produce-records: topicB-0
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k1, value=v1, timestamp=null)
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k2, value=v2, timestamp=null)
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k3, value=v3, timestamp=null)
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k4, value=v4, timestamp=null)
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k5, value=v5, timestamp=null)
    Segment[partition=topicB-0 offloaded-by-broker-id=0 base-offset=0 record-count=2]
    Segment[partition=topicB-0 offloaded-by-broker-id=0 base-offset=2 record-count=2]

[SUCCESS] (5) bounce-broker: 0

[SUCCESS] (6) consume-action:
  topic-partition = topicA-0
  fetch-offset = 1
  expected-record-count = 2
  expected-record-from-tiered-storage = 1

[SUCCESS] (7) consume-action:
  topic-partition = topicB-0
  fetch-offset = 1
  expected-record-count = 4
  expected-record-from-tiered-storage = 3

Content of local tiered storage:

Broker IDFile                                       | Offsets |      Records 
----------------------------------------------------------------------------------------
topicB-0                                            |         |              
    00000000000000000000-ZYHNDiMkQAuKx_Hm7yettA.log |       0 |     (k1, v1) 
                                                    |       1 |     (k2, v2) 
                                                    |         |              
    00000000000000000002-404zVl_XTuCyOEBrnEVxvQ.log |       2 |     (k3, v3) 
                                                    |       3 |     (k4, v4) 
                                                    |         |              
topicA-0                                            |         |              
    00000000000000000001-dwVJFEg8SmqzcKuJrD0C8w.log |       1 |     (k2, v2) 
                                                    |         |              
    00000000000000000000-dYoeCs5FSKSEekm7aBq3aQ.log |       0 |     (k1, v1) 
                                                    |         |              
```

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java
Comment (satishd) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java:130]: Can you remove prefix (A) in the docs of this class.

Comment (satishd) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java:129]: Please remove (A) from test documentation in this class. 
For ex:
```
Test Cases (A):
...
(A.1)
....
(A.2)
....
(A.3) Stops and restarts the broker.
...
```

**Review by satishd — COMMENTED** [review]
Thanks @kamalcph for the PR, overall LGTM, left a minor comment on the exising code.

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for addressing the review comments. LGTM.

File: storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java
Comment (showuon) [inline storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:257]: Will this cause we init clients twice when 1st start up? 
If we already reinit clients in `bounce` method, why should we reinit clients here?

**Review by showuon — COMMENTED** [review]
Overall LGTM! Thanks for the PR. Questions:
1. So, the solution we're going to adopt to fix the problem is only update log start offset in remote-log-segment deleter threads on becoming the leader. Is that right?
2. If so, I guess in the next patch, we'll add logics to handle the 2 cases raised in JIRA, is that right? 

Thank you.

File: storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java
Comment (kamalcph) [reply storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:257]: On broker startup, this method won't be invoked. Assume that the user does `stop` and `start` instead of `bounce`, then we may have to reinitialise the clients.

**Comment by kamalcph** [comment]
> Overall LGTM! Thanks for the PR. Questions:
> 
> 1. So, the solution we're going to adopt to fix the problem is only update log start offset in remote-log-segment deleter threads on becoming the leader. Is that right?
> 2. If so, I guess in the next patch, we'll add logics to handle the 2 cases raised in JIRA, is that right?
> 
> Thank you.

yes, case-1 will be handled by this approach. For case-2, we may have to check for end-to-end consistency and comeback on this.

File: storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java
Comment (showuon) [reply storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:257]: OK, I see.

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the fix.

**Comment by showuon** [comment]
Looks like there are problems while running CI. Jenkins is so slow now. Will try to check it later.

File: core/src/main/scala/kafka/log/UnifiedLog.scala
Comment (clolov) [inline core/src/main/scala/kafka/log/UnifiedLog.scala:1880]: Sorry it is probably obvious but what is the purpose of this refactor? Why can't you keep the logic in `remoteLogEnabled`?

**Comment by showuon** [comment]
Still cannot open this page: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14301

**Comment by clolov** [comment]
Can you pause merging this until I finish my review? I have a few questions about this change which I would like to clear up before it makes it in trunk?

**Comment by showuon** [comment]
> Can you pause merging this until I finish my review? I have a few questions about this change which I would like to clear up before it makes it in trunk?

@clolov , Got it! Will wait until getting your approval.

**Review by clolov — COMMENTED** [review]
Okay, thank you for the pull request, but can you go into a bit more detail as to what it addresses? As far as I can tell it fixes a test case, but it doesn't address either of the two cases specified in https://issues.apache.org/jira/browse/KAFKA-15351? In other words, I do not understand why we are ensuring that the log-start-offset is not updated to the local-log-start-offset when remote storage is enabled and why that is the correct approach? As far as I understood from the description of case 1 in the JIRA ticket we want to reconcile a stale log start offset from a replica with the latest global log start offset which I believe is only available via inspecting tiered storage?

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java
Comment (clolov) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java]: Is this true? Aren't there 2 segments in tiered storage for topic B?

File: core/src/main/scala/kafka/log/UnifiedLog.scala
Comment (kamalcph) [reply core/src/main/scala/kafka/log/UnifiedLog.scala:1880]: We want to know whether the remote log is enabled or not for the topic before the creation of `UnifiedLog` object. So, moved the logic to the companion object.

**Comment by kamalcph** [comment]
> Okay, thank you for the pull request, but can you go into a bit more detail as to what it addresses? As far as I can tell it fixes a test case, but it doesn't address either of the two cases specified in https://issues.apache.org/jira/browse/KAFKA-15351? In other words, I do not understand why we are ensuring that the log-start-offset is not updated to the local-log-start-offset when remote storage is enabled and why that is the correct approach? As far as I understood from the description of case 1 in the JIRA ticket we want to reconcile a stale log start offset from a replica with the latest global log start offset which I believe is only available via inspecting tiered storage?

Added the case-3 in the KAFKA-15351 ticket. This patch addresses only the case-3. We can follow-up the rest of the cases in a separate PR.

**Comment by clolov** [comment]
Okay, but I still do not understand, what is preventing the value in the log-start-checkpoint-offset from being stale as well? Or is the idea here that we will start with the log-start-*-offset which we knew was correct at some point in time and then we will contact tiered storage to try and figure out what later value it might have moved to? If so, does it matter which offset we start from as long as we have the logic for checking what is in tiered storage?

**Comment by kamalcph** [comment]
> Okay, but I still do not understand, what is preventing the value in the log-start-checkpoint-offset from being stale as well? Or is the idea here that we will start with the log-start-*-offset which we knew was correct at some point in time and then we will contact tiered storage to try and figure out what later value it might have moved to? If so, does it matter which offset we start from as long as we have the logic for checking what is in tiered storage?

yes, correct. The log-start-offset will be updated to the correct value eventually when the rlm-task updates the log-start-offset after becoming leader (patch needs to be opened). 

> If so, does it matter which offset we start from as long as we have the logic for checking what is in tiered storage?

If the log-start-offset moves ahead, then currently remote-log-segments are deleted due to breach by log-start-offset. Once the rlm-task updates the log-start-offset after becoming leader, then those remote-log-segments won't be removed. But, the log-start-offset should increase monotonically, it should not go forward and backward (at-least within a node).

**Comment by clolov** [comment]
Got it, okay, then this makes sense to me, thank you for the explanation!

**Review by clolov — APPROVED** [review]

**Review by divijvaidya — APPROVED** [review]

**Comment by showuon** [comment]
@kamalcph , looks like there are some end to end authorization tests failed due to this change. Could you help fix them? Thanks.
https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14301/7/#showFailuresLink

**Comment by kamalcph** [comment]
> @kamalcph , looks like there are some end to end authorization tests failed due to this change. Could you help fix them? Thanks. https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14301/7/#showFailuresLink

checking the test failures.

**Comment by kamalcph** [comment]
> @kamalcph , looks like there are some end to end authorization tests failed due to this change. Could you help fix them? Thanks. https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14301/7/#showFailuresLink

Test failures are fixed. Please take another look!

**Comment by satishd** [comment]
Thanks @kamalcph for fixing the test failures. The test failures with the latest job are unrelated to this change. Merging to trunk and 3.6.

File: core/src/main/scala/kafka/log/LogLoader.scala
Comment (junrao) [inline core/src/main/scala/kafka/log/LogLoader.scala:82]: Could we add the new param to the javadoc?

Comment (kamalcph) [reply core/src/main/scala/kafka/log/LogLoader.scala:82]: Opened a PR to address it: https://github.com/apache/kafka/pull/15179. PTAL

**Pull request merged**
